gr-uhd/lib/rfnoc_siggen_impl.cc

```diff
@@ -0,0 +1,158 @@
+void rfnoc_siggen_impl::set_amplitude(const double amplitude, const size_t chan = 0)
+{
+    d_siggen_ref->set_amplitude(amplitude, chan);
+}
+
+double rfnoc_siggen_impl::get_amplitude(const size_t chan = 0)
+{
+    return d_siggen_ref->get_amplitude(chan);
+}
+
+void rfnoc_siggen_impl::set_constant(const std::complex<double> constant,
+                                     const size_t chan = 0)
+{
+    d_siggen_ref->set_constant(constant, chan);
+}
+
+std::complex<double> rfnoc_siggen_impl::get_constant(const size_t chan = 0)
+{
+    return d_siggen_ref->get_constant(chan);
+}
+
+void rfnoc_siggen_impl::set_enable(const bool enable, const size_t chan = 0)
+{
+    d_siggen_ref->set_enable(enable, chan);
+}
+
+bool rfnoc_siggen_impl::get_enable(const size_t chan = 0)
+{
+    return d_siggen_ref->get_enable(chan);
+}
+
+void rfnoc_siggen_impl::set_sine_phase_increment(const double phase_inc,
+                                                 const size_t chan = 0)
+{
+    d_siggen_ref->set_sine_phase_increment(phase_inc, chan);
+}
+
+double rfnoc_siggen_impl::get_sine_phase_increment(const size_t chan = 0)
+{
+    return d_siggen_ref->get_sine_phase_increment(chan);
+}
+
+void rfnoc_siggen_impl::set_sine_frequency(const double frequency,
+                                           const double sample_rate,
+                                           const size_t chan = 0)
+{
+    d_siggen_ref->set_sine_frequency(frequency, sample_rate, chan);
+}
+
+void rfnoc_siggen_impl::set_waveform(const rfnoc_siggen::siggen_waveform type,
+                                     const size_t chan = 0)
+{
+    d_siggen_ref->set_waveform(waveform_type_gr2u(type), chan);
+}
+
+void rfnoc_siggen_impl::set_waveform(const std::string& type, const size_t chan = 0)
+{
+    if (type == "NOISE") {
+        d_siggen_ref->set_waveform(::uhd::rfnoc::siggen_waveform::NOISE, chan);
+    } else if (type == "CONSTANT") {
+        d_siggen_ref->set_waveform(::uhd::rfnoc::siggen_waveform::CONSTANT, chan);
+    } else if (type == "SINE_WAVE") {
+        d_siggen_ref->set_waveform(::uhd::rfnoc::siggen_waveform::SINE_WAVE, chan);
+    } else {
+        throw std::runtime_error("SigGen waveform type (" + type +
+                                 ") is not [NOISE, CONSTANT, SINE_WAVE]");
+    }
+}
+
+rfnoc_siggen::siggen_waveform rfnoc_siggen_impl::get_waveform(const size_t chan)
+{
+    return waveform_type_u2gr(d_siggen_ref->get_waveform(chan));
+}
+
+std::string rfnoc_siggen_impl::get_waveform_string(const size_t chan)
+{
+    const auto type = get_waveform(chan);
+    switch (type) {
+    case siggen_waveform::NOISE:
+        return "NOISE";
+    case siggen_waveform::CONSTANT:
+        return "CONSTANT";
+    case siggen_waveform::SINE_WAVE:
+        return "SINE_WAVE";
+    default:
+        throw std::runtime_error("SigGen waveform type enumeration (" +
+                                 std::to_string(static_cast<uint32_t>(type)) +
```

Comment:
here' using `fmt::format("SigGen waveform type enumeration ({:d}) is not [NOISE, CONSTANT, SINE_WAVE]", static_cast<uint32_t>(type));` might actually be easier to read.  (though we both probably agree, this is a scoped enum. If we *don't* have a `default` case in a switch over the same, then our compiler will warn us if we don't cover all enum values, and the return value of `get_waveform` actually guarantees there's none not-in-enum values. Still, having this default case makes future extensions less bug-prone, so :+1: )